src/celeritas/optical/CerenkovPreGenerator.hh

```diff
@@ -0,0 +1,139 @@
+    inline CELER_FUNCTION
+    CerenkovPreGenerator(ParticleTrackView const& particle_view,
+                         NativeCRef<OpticalPropertyData> const& properties,
+                         NativeCRef<CerenkovData> const& shared,
+                         OpticalMaterialId mat_id,
+                         OpticalStepCollectorData const& optical_step);
+
+    // Return a populated distribution data for the Cerenkov Generator
+    template<class Generator>
+    inline CELER_FUNCTION OpticalDistributionData operator()(Generator& rng);
+
+  private:
+    units::ElementaryCharge charge_;
+    OpticalMaterialId mat_id_;
+    OpticalStepCollectorData step_collector_data_;
+    real_type num_photons_per_len_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from particle, material, and optical step collection data.
+ */
+CELER_FUNCTION CerenkovPreGenerator::CerenkovPreGenerator(
+    ParticleTrackView const& particle_view,
+    NativeCRef<OpticalPropertyData> const& properties,
+    NativeCRef<CerenkovData> const& shared,
+    OpticalMaterialId mat_id,
+    OpticalStepCollectorData const& optical_step)
+    : charge_(particle_view.charge())
+    , mat_id_(mat_id)
+    , step_collector_data_(optical_step)
+{
+    CELER_EXPECT(charge_.value() != 0);
+    CELER_EXPECT(mat_id_);
+    CELER_EXPECT(step_collector_data_);
+
+    auto const& pre = step_collector_data_.points[StepPoint::pre];
+    auto const& post = step_collector_data_.points[StepPoint::post];
+    units::LightSpeed beta(real_type{0.5}
+                           * (pre.speed.value() + post.speed.value()));
+
+    CerenkovDndxCalculator dndx_calculator(
+        properties, shared, mat_id_, charge_);
+    num_photons_per_len_ = dndx_calculator(beta);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return an \c OpticalDistributionData object. If no photons are sampled, an
+ * empty object is returned and can be verified via its own operator bool.
+ *
+ * The number of photons is sampled from a Poisson distribution with a mean
+ * \f[
+   \langle n \rangle = \ell_\text{step} \frac{dN}{dx}
+ * \f]
+ * where \f$ \ell_\text{step} \f$ is the step length.
+ */
+template<class Generator>
+CELER_FUNCTION OpticalDistributionData
+CerenkovPreGenerator::operator()(Generator& rng)
+{
+    if (num_photons_per_len_ == 0)
+    {
+        return {};
+    }
+
+    // Sample number of photons from a Poisson distribution
+    auto sampled_num_photons = PoissonDistribution<real_type>(
+        num_photons_per_len_ * step_collector_data_.step_length)(rng);
+
+    if (sampled_num_photons == 0)
+    {
+        // Not an optical material or this step is below production threshold
+        return {};
+    }
```

Comment:
Is this special case necessary? I'd just put the `data.{charge ... material}` inside an `if (data.num_photons > 0)` block.

Reply:
Similar logical comment for the comment line - Cerenkov pre-generator should be only called if the material where the step is, is an optical material.

Reply:
Fair assumption. Guess I could update the comment to state it will return a false `OpticalDistributionData` object if step is below threshold, as we may not check the threshold before calling the pre-generator (just let it tell us that the number of sampled photons on that step is zero).